scanpy/tools/_density.py

```diff
@@ -0,0 +1,139 @@
+"""Calculate density of cells in embeddings
+"""
+
+import numpy as np
+from scipy.stats import gaussian_kde
+from anndata import AnnData
+
+from .. import logging as logg
+
+
+def _calc_density(
+        x: np.ndarray,
+        y: np.ndarray):
+    """
+    Function to calculate the density of cells in an embedding.
+    """    
+
+    # Calculate the point density
+    xy = np.vstack([x,y])
+    z = gaussian_kde(xy)(xy)
+
+    min_z = np.min(z)
+    max_z = np.max(z)
+
+    # Scale between 0 and 1
+    scaled_z = (z-min_z)/(max_z-min_z)
+
+    return(scaled_z)
+
+
+def density(
+        adata: AnnData,
+        embedding: str,
+        groupby: str = None,
+        key_added: str = None):
+    """Calculate the density of cells in an embedding (per condition)
+
+    Gaussian kernel density estimation is used to calculate the density of
+    cells in an embedded space. This can be performed per category over a
+    categorical cell annotation. The cell density can be plotted using the 
+    `sc.pl.density()` function.
+
+    This function was written by Sophie Tritschler and implemented into
+    Scanpy by Malte Luecken.
+    
+    Parameters
+    ----------
+    adata : :class:`~anndata.AnnData`
+        The annotated data matrix.
+    embedding : `str`
+        The embedding over which the density will be calculated. This must
+        be one of:
+        'umap' : UMAP
+        'dm' : Diffusion map
+        'pca' : PCA
+        'tsne' : t-SNE
+        'fa' : Force-directed graph layout by Force Atlas 2
```

Comment:
I think this is same as the `basis` option we have in other plotting functions. There is no need to hard-code basis, it shouldn't be limited to predefined basis as in here. We can use spatial basis, dca or scvi basis etc., if they're stored in obsm.

You can use the same processing here: https://github.com/theislab/scanpy/blob/master/scanpy/plotting/_anndata.py#L300

There is also a trick to skip the useless dimension of diffmap there.

Reply:
This is indeed the same... I should rename the argument basis. And yes, the adaptation for diffmap is very helpful.

In terms of not hard-coding I guess I can just throw an error to saw that `adata.obsm['X_'+basis]` does not exist then?

Reply:
Sounds good.